File: IMSS/Codigos/3.py
import pandas as pd
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import confusion_matrix, accuracy_score
import matplotlib.pyplot as plt
import seaborn as sns

# 1. Cargar dataset
data_path = r"C:\Users\Zainos\Downloads\Practica 3\Corregido\2_datos_numericos.csv"
df = pd.read_csv(data_path)

target = "FE_FinalNR"
X = df.drop(columns=[target])
y = df[target]

# 2. Codificación dummies para nominales
nominales = X.select_dtypes(include=["object"]).columns.tolist()
X = pd.get_dummies(X, columns=nominales, drop_first=True)

# 3. Split manual por clase
train_list = []
val_list = []
test_list = []

train_frac = 0.8
val_frac_within_train = 0.2

distribucion = []

for clase in sorted(y.unique()):
    idx = y[y == clase].index
    df_clase = X.loc[idx].copy()
    y_clase = y.loc[idx].copy()
    
    df_clase = df_clase.sample(frac=1, random_state=42)
    y_clase = y_clase.loc[df_clase.index]
    
    n_total = len(df_clase)
    n_train_full = int(n_total * train_frac)
    n_test = n_total - n_train_full
    n_val = int(n_train_full * val_frac_within_train)
    n_train = n_train_full - n_val
    
    train_list.append((df_clase.iloc[:n_train], y_clase.iloc[:n_train]))
    val_list.append((df_clase.iloc[n_train:n_train+n_val], y_clase.iloc[n_train:n_train+n_val]))
    test_list.append((df_clase.iloc[n_train+n_val:], y_clase.iloc[n_train+n_val:]))
    
    distribucion.append({
        "Clase": clase,
        "Train": n_train,
        "Validation": n_val,
        "Test": n_test
    })

X_train = pd.concat([t[0] for t in train_list])
y_train = pd.concat([t[1] for t in train_list])

X_val = pd.concat([v[0] for v in val_list])
y_val = pd.concat([v[1] for v in val_list])

X_test = pd.concat([t[0] for t in test_list])
y_test = pd.concat([t[1] for t in test_list])

# 4. Mostrar distribución por clase
df_distribucion = pd.DataFrame(distribucion)
print("Distribución por clase en cada conjunto:")
print(df_distribucion)

# 5. Entrenar árbol de decisión
clf = DecisionTreeClassifier(random_state=42, max_depth=5)
clf.fit(X_train, y_train)

# 6. Predicción y evaluación
y_pred = clf.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
acc = accuracy_score(y_test, y_pred)

print("\nMatriz de confusión:\n", cm)
print(f"\nPrecisión: {acc:.3f}")

# 7. Importancia de variables
importancias = pd.DataFrame({
    "Variable": X_train.columns,
    "Importancia": clf.feature_importances_
}).sort_values(by="Importancia", ascending=False)

print("\nVariables más importantes:\n", importancias.head(10))

# 8. Visualización del árbol como imagen
labels = [str(c) for c in sorted(y.unique())]

plt.figure(figsize=(50,20))
plot_tree(clf, feature_names=X_train.columns, class_names=labels, filled=True, fontsize=12)
ruta_arbol = r"C:\Users\Zainos\Downloads\Practica 3\Corregido\arbol_decision.png"
plt.savefig(ruta_arbol, dpi=300, bbox_inches='tight')
plt.close()
print(f"Árbol de decisión guardado en: {ruta_arbol}")

# 9. Visualización exploratoria
plt.figure(figsize=(10,5))
sns.countplot(x=y_test)
plt.title("Distribución de FE_FinalNR en el conjunto de prueba")
plt.xlabel("FE_FinalNR")
plt.ylabel("Cantidad de observaciones")
plt.show()

if 'edad' in X_test.columns:
    plt.figure(figsize=(10,6))
    sns.boxplot(x=y_test, y=X_test['edad'])
    plt.title("Boxplot de edad por FE_FinalNR")
    plt.xlabel("FE_FinalNR")
    plt.ylabel("Edad")
    plt.show()

if 'entidad' in df.columns:
    plt.figure(figsize=(12,6))
    sns.countplot(x=df['entidad'], hue=df[target])
    plt.title("Distribución de FE_FinalNR por entidad")
    plt.xlabel("Entidad")
    plt.ylabel("Cantidad de observaciones")
    plt.xticks(rotation=90)
    plt.legend(title='FE_FinalNR')
    plt.show()

plt.figure(figsize=(15,10))
sns.heatmap(df.select_dtypes(include='number').corr(), cmap="coolwarm", annot=False, cbar=True)
plt.title("Mapa de correlación de variables numéricas")
plt.show()

plt.figure(figsize=(8, 6))
sns.boxplot(x="cal1", y="edad", data=df)
plt.title("Edad vs Nivel de satisfacción")
plt.show()

plt.figure(figsize=(8, 6))
sns.barplot(x="sexo", y="cal1", data=df, estimator="mean", ci=None)
plt.title("Promedio de satisfacción por sexo")
plt.show()

df["edad"].hist(bins=20)
plt.title("Distribución de la edad")
plt.xlabel("Edad")
plt.ylabel("Frecuencia")
plt.show()

File: IMSS/Codigos/3.R
# 3.R — Árbol de decisión (estratificado por clase), métricas, importancias y visualizaciones

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(ggplot2)
  library(rpart)
  library(rpart.plot)
  library(reshape2)
})

# -----------------------------
# Rutas
# -----------------------------
data_path  <- "C:/Users/Zainos/Downloads/Practica 3/Corregido/2_datos_numericos.csv"
ruta_arbol <- "C:/Users/Zainos/Downloads/Practica 3/Corregido/arbol_decision.png"

# -----------------------------
# Cargar dataset
# -----------------------------
df <- readr::read_csv(data_path, show_col_types = FALSE)
target <- "FE_FinalNR"

stopifnot(target %in% names(df))
X <- df %>% dplyr::select(-all_of(target))
y <- df[[target]]

# -----------------------------
# Dummies para nominales (si hubiera columnas no numéricas)
# -----------------------------
# Si todo ya está numérico (por el script 2), esto no cambia nada.
X_model <- if (any(!vapply(X, is.numeric, logical(1)))) {
  mm <- model.matrix(~ . - 1, data = mutate(X, across(!where(is.numeric), as.factor)))
  as.data.frame(mm)
} else {
  as.data.frame(X)
}

# -----------------------------
# Split estratificado por clase (80% train, de eso 20% val, resto test)
# -----------------------------
set.seed(42)
clases <- sort(unique(y))
train_list <- list(); val_list <- list(); test_list <- list()
train_frac <- 0.8
val_frac_within_train <- 0.2

distribucion <- list()

for (cl in clases) {
  idx <- which(y == cl)
  idx <- sample(idx, length(idx))  # shuffle
  
  n_total <- length(idx)
  n_train_full <- floor(n_total * train_frac)
  n_test <- n_total - n_train_full
  n_val <- floor(n_train_full * val_frac_within_train)
  n_train <- n_train_full - n_val
  
  tr_idx  <- idx[seq_len(n_train)]
  va_idx  <- idx[seq(n_train + 1, n_train + n_val)]
  te_idx  <- idx[(n_train + n_val + 1):n_total]
  
  train_list[[as.character(cl)]] <- list(X = X_model[tr_idx, , drop = FALSE], y = y[tr_idx])
  val_list[[as.character(cl)]]   <- list(X = X_model[va_idx, , drop = FALSE], y = y[va_idx])
  test_list[[as.character(cl)]]  <- list(X = X_model[te_idx, , drop = FALSE], y = y[te_idx])
  
  distribucion[[as.character(cl)]] <- data.frame(
    Clase = cl, Train = n_train, Validation = n_val, Test = n_test
  )
}

X_train <- do.call(rbind, lapply(train_list, `[[`, "X"))
y_train <- unlist(lapply(train_list, `[[`, "y"))

X_val   <- do.call(rbind, lapply(val_list, `[[`, "X"))
y_val   <- unlist(lapply(val_list, `[[`, "y"))

X_test  <- do.call(rbind, lapply(test_list, `[[`, "X"))
y_test  <- unlist(lapply(test_list, `[[`, "y"))

df_distribucion <- dplyr::bind_rows(distribucion)
cat("Distribución por clase en cada conjunto:\n"); print(df_distribucion)

# -----------------------------
# Entrenar árbol de decisión
# -----------------------------
train_df <- cbind(X_train, .y = as.factor(y_train))
fit <- rpart(.y ~ ., data = train_df,
             method = "class",
             control = rpart.control(maxdepth = 5))

# -----------------------------
# Predicción y evaluación
# -----------------------------
pred <- predict(fit, newdata = X_test, type = "class")
cm <- table(Real = y_test, Pred = pred)
acc <- sum(diag(cm)) / sum(cm)
cat("\nMatriz de confusión:\n"); print(cm)
cat(sprintf("\nPrecisión: %.3f\n", acc))

# -----------------------------
# Importancia de variables
# -----------------------------
if (!is.null(fit$variable.importance)) {
  importancias <- data.frame(
    Variable = names(fit$variable.importance),
    Importancia = as.numeric(fit$variable.importance),
    row.names = NULL
  ) %>% arrange(desc(Importancia))
  cat("\nVariables más importantes (top 10):\n")
  print(head(importancias, 10))
} else {
  cat("\nEl modelo no reporta importancias.\n")
}

# -----------------------------
# Visualización del árbol como imagen
# -----------------------------
png(filename = ruta_arbol, width = 2000, height = 1200, res = 150)
rpart.plot::rpart.plot(fit, type = 2, extra = 106, under = TRUE, faclen = 0)
dev.off()
cat(sprintf("Árbol de decisión guardado en: %s\n", ruta_arbol))

# -----------------------------
# Visualizaciones exploratorias (opcionales, si existen columnas)
# -----------------------------
# Distribución de FE_FinalNR en test
ggplot(data.frame(FE_FinalNR = y_test), aes(x = FE_FinalNR)) +
  geom_bar() + theme_minimal() +
  ggtitle("Distribución de FE_FinalNR en el conjunto de prueba") +
  xlab("FE_FinalNR") + ylab("Cantidad")

# Boxplot de edad por FE_FinalNR (si existe 'edad' en X_test o df)
if ("edad" %in% names(df)) {
  ggplot(df, aes(x = .data[[target]], y = .data[["edad"]])) +
    geom_boxplot() + theme_minimal() +
    ggtitle("Edad vs Nivel de satisfacción") +
    xlab(target) + ylab("Edad")
}

# Distribución por entidad (si existe 'entidad')
if ("entidad" %in% names(df)) {
  ggplot(df, aes(x = entidad, fill = .data[[target]])) +
    geom_bar(position = "stack") + theme_minimal() +
    ggtitle("Distribución de FE_FinalNR por entidad") +
    xlab("Entidad") + ylab("Cantidad de observaciones") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
}

# Mapa de correlación de variables numéricas
num_df <- dplyr::select(df, where(is.numeric))
if (ncol(num_df) >= 2) {
  corr <- stats::cor(num_df, use = "pairwise.complete.obs")
  molten <- reshape2::melt(corr)
  ggplot(molten, aes(Var1, Var2, fill = value)) +
    geom_tile() +
    scale_fill_gradient2(limits = c(-1, 1)) +
    theme_minimal() +
    ggtitle("Mapa de correlación de variables numéricas") +
    xlab(NULL) + ylab(NULL)
}

# Ejemplo: boxplot edad vs cal1 (si existen)
if (all(c("edad", "cal1") %in% names(df))) {
  ggplot(df, aes(x = as.factor(cal1), y = edad)) +
    geom_boxplot() + theme_minimal() +
    ggtitle("Edad vs Nivel de satisfacción") +
    xlab("cal1") + ylab("Edad")
}

# Promedio de satisfacción por sexo (si existen)
if (all(c("sexo", "cal1") %in% names(df))) {
  ggplot(df, aes(x = as.factor(sexo), y = cal1)) +
    stat_summary(fun = mean, geom = "bar") +
    theme_minimal() +
    ggtitle("Promedio de satisfacción por sexo") +
    xlab("sexo") + ylab("Promedio cal1")
}

# Histograma de edad (si existe)
if ("edad" %in% names(df)) {
  ggplot(df, aes(x = edad)) +
    geom_histogram(bins = 20) +
    theme_minimal() +
    ggtitle("Distribución de la edad") +
    xlab("Edad") + ylab("Frecuencia")
}
